.github/workflows/ci.yml

```diff
@@ -5,134 +5,237 @@ concurrency:
-          key: ${{ runner.os }}-lint-modules-${{ secrets.CACHE_VERSION }}-${{ hashFiles('**/yarn.lock') }}
-      - name: Install dependencies
+          path: |
+            **/node_modules
+          key: ${{ runner.os }}-dependencies-${{ secrets.CACHE_VERSION }}-${{ hashFiles('yarn.lock') }}
+      - name: Install dependencies and build monorepo
         run: yarn install --frozen-lockfile
-      - name: Run linter
-        run: yarn run lint
+      - name: Install ${{ matrix.browser }} via Playwright
+        if: startsWith(matrix.browser, 'Webkit')
+        # The @L will convert 'Webkit' into lowercase 'webkit' which is the name Playwright expects.
+        run: yarn playwright install --with-deps ${BROWSER_NAME@L}
+        env:
+          BROWSER_NAME: ${{ matrix.browser }}
+      - name: Run browser tests
+        uses: nick-fields/retry@v3
+        with:
+          timeout_minutes: 2
+          max_attempts: 3
+          # The tests will run on 'WebkitHeadless', 'FirefoxHeadless' and 'ChromeHeadless'
+          command: yarn run test:browser --browsers ${{ matrix.browser }}Headless
 
-  browser:
+  coveralls:
+    name: Coverage check
+    needs: test
     runs-on: ubuntu-latest
     steps:
-      - name: Use Node.js
+      - name: Consolidate test coverage from different jobs
+        uses: coverallsapp/github-action@v2
+        with:
+          github-token: ${{ secrets.github_token }}
+          parallel-finished: true
+
+  browserbuilds:
+    name: Browser builds
+    needs:
+      - test
+      - spectest
+      - browsertest
+    runs-on: ubuntu-latest
+    steps:
+      - name: Ensure line endings are consistent
+        run: git config --global core.autocrlf input
+      - name: Use Node.js ${{ env.DEFAULT_NODE_VERSION }}
         uses: actions/setup-node@v4
         with:
-          node-version: 18.x
+          node-version: ${{ env.DEFAULT_NODE_VERSION }}
       - name: Check out repository
         uses: actions/checkout@v4
-      - name: Load cache
+      - name: Load dependency cache
         uses: actions/cache@v4
         with:
-          path: '**/node_modules'
-          key: ${{ runner.os }}-browser-modules-${{ secrets.CACHE_VERSION }}-${{ hashFiles('**/yarn.lock') }}
-      - name: Install dependencies
+          path: |
+            **/node_modules
+          key: ${{ runner.os }}-dependencies-${{ secrets.CACHE_VERSION }}-${{ hashFiles('yarn.lock') }}
+      - name: Install dependencies and build monorepo
         run: yarn install --frozen-lockfile
-      - name: Build for browser
-        run: npx lerna run browser --concurrency 1
-      - name: Deploy browser build
+      - name: Build browser bundles
+        run: yarn run build:browser
+      - name: Deploy browser builds
         if: startsWith(github.ref, 'refs/heads/master') || startsWith(github.ref, 'refs/tags/')
         run: ./.github/deploy_browser.sh
         env:
           GH_TOKEN: ${{ secrets.PAT }}
 
-  docker:
+  dockerimages:
```

Comment:
With the docker and browser phases being only run after the test stuff, total workflow duration increases from 12m 59s to 14m 48s.
But it may be better like this, so we ensure that stuff is only deployed after we're certain all test stuff pass.